src/procedural-macros.md

```diff
@@ -1,24 +1,546 @@
-Procedural macros can be used to implement custom [derive] on your own
-types. See [the book][procedural macros] for a tutorial.
+Procedural macros currently come in one of three flavors:
 
-Procedural macros involve a few different parts of the language and its
-standard libraries. First is the `proc_macro` crate, included with Rust,
-that defines an interface for building a procedural macro. The
-`#[proc_macro_derive(Foo)]` attribute is used to mark the deriving
-function. This function must have the type signature:
+* Custom macros - `my_macro!(...)`
+* Custom derive - `#[derive(MyTrait)]`
+* Custom attributes - `#[my_attribute]`
+
+Procedural macros allow you to run code at compile time that operates over Rust
+syntax, both consuming and producing Rust syntax. You can sort of think of
+procedural macros as Rust functions from an AST to another AST.
+
+### Crates and procedural macros
+
+All procedural macros in Rust are compiled as a crate. Procedural macro crates
+are defined with Cargo via the `proc-macro` key in your manfiest:
+
+```toml
+[lib]
+proc-macro = true
+```
+
+Procedural macros are always compiled with the same target as the compiler
+itself. For example if you execute `cargo build --target
+arm-unknown-linux-gnueabi` then procedural macros will not be compiled for ARM,
+but rather your build computer (for example `x86_64-unknown-linux-gnu`).
+
+Procedural macro crates are not currently allowed to export any items except
+procedural macros (we'll see how to define these in a bit). For example this
+crate is not allowed:
+
+```rust
+pub fn foo() {}
+```
+
+because the `foo` function is not a procedural macro. Procedural macros are
+loaded dynamically by the compiler when they are needed during compilation.
+Cargo will naturally make procedural macro crates available to crates which
+depend on them, or you can use the `--extern` argument.
+
+### The `proc_macro` crate
+
+Procedural macro crates almost always will link to the in-tree `proc_macro`
+crate. The `proc_macro` crate is a compiler-provided crate which provides
+facilities to working with the types of each procedural macro function. You can
+learn more about this crate by exploring [the documentation][pm-dox].
+
+[pm-dox]: https://doc.rust-lang.org/stable/proc_macro/
+
+Linking to the `proc_macro` crate can currently be done with:
 
 ```rust,ignore
-use proc_macro::TokenStream;
+extern crate proc_macro;
+```
+
+In the 2018 edition, however, this statement will not be necessary.
+
+### The `TokenStream` Type
+
+One aspect you may notice about the `proc_macro` crate is that it doesn't
+contain any AST items! Instead, it primarily contains a `TokenStream` type.
+Procedural macros in Rust operate over *token streams* instead of AST nodes,
+which is a far more stable interface over time for both the compiler and for
+procedural macros to target.
+
+A *token stream* is roughly equivalent to `Vec<TokenTree>` where a `TokenTree`
```

Comment:
Somewhere it would be good to mention the key difference between `TokenStream` and `Vec<TokenTree>` -- that `TokenStream` is cheap to clone. Though maybe this belongs in rustdoc of the `TokenStream` type.